src/target/riscv/riscv-013.c

```diff
@@ -4341,77 +4326,127 @@ static int read_memory_progbuf_inner_one(struct target *target,
+
 	uint64_t mstatus = 0;
 	uint64_t mstatus_old = 0;
 	if (modify_privilege(target, &mstatus, &mstatus_old) != ERROR_OK)
 		return MEM_ACCESS_FAILED_PRIV_MOD_FAILED;
 
-	const bool mprven = riscv_virt2phys_mode_is_hw(target)
+	const bool mprven_required = riscv_virt2phys_mode_is_hw(target)
 			&& get_field(mstatus, MSTATUS_MPRV);
-	int result = (args.count == 1) ?
-		read_memory_progbuf_inner_one(target, args, mprven) :
-		read_memory_progbuf_inner(target, args, mprven);
 
-	if (mstatus != mstatus_old &&
-			register_write_direct(target, GDB_REGNO_MSTATUS, mstatus_old) != ERROR_OK)
-		return MEM_ACCESS_FAILED;
+	result = is_read ? read_memory_progbuf(target, args, mprven_required) :
+			write_memory_progbuf(target, args, mprven_required);
 
-	return (result == ERROR_OK) ? MEM_ACCESS_OK : MEM_ACCESS_FAILED;
+	if (mstatus != mstatus_old && register_write_direct(target,
+			GDB_REGNO_MSTATUS, mstatus_old) != ERROR_OK)
+		return MEM_ACCESS_FAILED_REG_WRITE_FAILED;
+
+	return result;
 }
 
+static int
+write_memory_bus_v0(struct target *target, const riscv_mem_access_args_t args);
+static int
+write_memory_bus_v1(struct target *target, const riscv_mem_access_args_t args);
+
 static mem_access_result_t
-read_memory_sysbus(struct target *target, const riscv_mem_access_args_t args)
+access_memory_sysbus(struct target *target, const riscv_mem_access_args_t args)
 {
-	assert(riscv_mem_access_is_read(args));
+	assert(riscv_mem_access_is_valid(args));
 
-	mem_access_result_t skip_reason = mem_should_skip_sysbus(target, args);
-	if (skip_reason != MEM_ACCESS_OK)
-		return skip_reason;
+	mem_access_result_t result = mem_should_skip_sysbus(target, args);
+	if (result != MEM_ACCESS_OK)
+		return result;
 
-	int ret = ERROR_FAIL;
-	uint64_t sbver = get_field(get_info(target)->sbcs, DM_SBCS_SBVERSION);
+	RISCV013_INFO(info);
+	int res = ERROR_FAIL;
+	const bool is_read = riscv_mem_access_is_read(args);
+	const uint64_t sbver = get_field(info->sbcs, DM_SBCS_SBVERSION);
 	if (sbver == 0)
-		ret = read_memory_bus_v0(target, args);
+		res = is_read ? read_memory_bus_v0(target, args) :
+				write_memory_bus_v0(target, args);
 	else if (sbver == 1)
-		ret = read_memory_bus_v1(target, args);
+		res = is_read ? read_memory_bus_v1(target, args) :
+				write_memory_bus_v1(target, args);
 	else
-		LOG_TARGET_ERROR(target,
-			"Unknown system bus version: %" PRIu64, sbver);
+		LOG_TARGET_ERROR(target, "Unknown system bus version: %" PRIu64, sbver);
 
-	return (ret == ERROR_OK) ? MEM_ACCESS_OK : MEM_ACCESS_FAILED;
+	return res == ERROR_OK ? MEM_ACCESS_OK : MEM_ACCESS_FAILED;
 }
 
-static int read_memory(struct target *target, const riscv_mem_access_args_t args)
+static mem_access_result_t
+access_memory_abstract(struct target *target, const riscv_mem_access_args_t args)
 {
-	assert(riscv_mem_access_is_read(args));
+	assert(riscv_mem_access_is_valid(args));
 
-	if (args.count == 0)
-		return ERROR_OK;
+	mem_access_result_t res = mem_should_skip_abstract(target, args);
```

Comment:
```suggestion
	mem_access_result_t skip_reason = mem_should_skip_abstract(target, args);
```

I would prefer if this kept it's old name.

Reply:
Thanks! Addressed